OpenRA.Mods.Common/Traits/Conditions/GrantConditionOnLayer.cs

```diff
@@ -0,0 +1,73 @@
+#region Copyright & License Information
+/*
+ * Copyright 2007-2018 The OpenRA Developers (see AUTHORS)
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation, either version 3 of
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+using System.Linq;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.Common.Traits
+{
+	public abstract class GrantConditionOnLayerInfo : ConditionalTraitInfo
+	{
+		[FieldLoader.Require]
+		[GrantedConditionReference]
+		[Desc("The condition to grant to self when changing to specific custom layer.")]
+		public readonly string Condition = null;
+	}
+
+	public abstract class GrantConditionOnLayer : ConditionalTrait<GrantConditionOnLayerInfo>, INotifyCustomLayerChanged
```

Comment:
Please declare this as:
```csharp
public abstract class GrantConditionOnLayer<InfoType> : ConditionalTrait<InfoType>, INotifyCustomLayerChanged where InfoType : ConditionalTraitInfo
```

and then change `GrantConditionOnLayerInfo` to `InfoType` in the constructor below.

This then allows you to declare the correct *Info type from the subclasses using e.g.:
```csharp
public class GrantConditionOnSubterraneanLayer : GrantConditionOnLayer<GrantConditionOnSubterraneanLayerInfo>, INotifyVisualPositionChanged
```
and you can then drop the `localInfo` declarations.